torch/_dynamo/__init__.py

```diff
@@ -112,3 +112,49 @@ def fn(a):
 def graph_break():
     """Force a graph break"""
     pass
+
+
+def forbid_in_graph(fn):
+    """
+    Customize which functions TorchDynamo will assert are not present while tracing.
```

Comment:
I don't think this really makes it clear. Maybe this extra text would help. "In particular, if you call a `forbid_in_graph` function inside a `torch.compile` region, this will result in a hard error and indicate this function call should be moved outside of the region." Which now makes me wonder, if you are in a torch.compile region but you happen to be in a frame that is not being compiled, should we raise anyway? This seems like a good idea in case we improve dynamo and now your frame is being compiled.

Reply:
We should only raise when we hit upon this value in tracing - there is lots of prod code that sets up the analogue of `if tracing / if not tracing` and uses it to safeguard regions from dynamo access or to change behavior during tracing. The prior art for this, graph breaks, only occur when you happen upon it. If a frame with a graph break in it is not being compiled, for whatever reason, we do not graph break (in export, we do not error). I think this should match.

Reply:
I added some tests clarifying this - lmk if they align with what you expect.

Reply:
No, I don't think the analogy to graph breaks is applicable. Graph breaks never cause code to raise an error; it's always semantics preserving to add them, no matter if they do something or not. But forbidden operators CAN cause code to raise an error. I am saying it is more uniform and easier to predict if they raise an error if they ALWAYS raise an error based on some simple condition (is torch.compile in my stack) as opposed to a complicated one (did I happen to be processing this frame or not.)